### PR TITLE
Make pygments optional for pyface.api

### LINKED ISSUE #655: Consider making pygments an optional dependency for pyface.api
`pygments` is only required for `CodeEditor` and `PythonShell` (qt). For use cases where neither of these objects are required, one still needs to have `pygments` in the environment in order to import anything from `pyface.api`.

Consider deferring this dependency until the said widgets are required.

### PR DESCRIPTION
**Pull request by kitchoi**
Closes #655

Depending on the project context, PythonEditor and PythonShell may never be needed, and yet importing `pyface.api` still requires `pygments`. This PR modifies `pyface.api` such that if import of PythonEditor and PythonShell fail because of missing pygments, it will carry on. It makes maintaining dependencies easier for projects that don't use PythonEditor/PythonShell.
 
This PR does not modify the `extras_requires` in setup.py. Pip-installing pyface with the Qt tag is still going to bring in pygments, see discussion in #655.

### CODE REVIEW

File: pyface/tests/test_api.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+import unittest
+
+
+class TestApi(unittest.TestCase):
+
+    def test_api_importable(self):
+        # make sure api is importable with the most minimal
+        # required dependencies, including in the absence of toolkit backends.
+        from pyface import api   # noqa: F401
```
Comment (kitchoi) [inline pyface/tests/test_api.py:21]: I verified this test failed when I have an environment with Qt but no pygments. Then the test passes after I made the change in api.py.

Ideally our CI matrix should cover more minimal setups...

**Review by corranwebster — CHANGES_REQUESTED** [review]
This will block usage of the `PythonEditor` and the `PythonShell` with wx if Pygments is not installed, even though the Wx backend doesn't use Pygments.

I think if you move this import:
https://github.com/enthought/pyface/blob/4bccc3e4bba7b7748025b21e08359f446a2abaa7/pyface/ui/qt4/code_editor/code_widget.py#L22
to here:
https://github.com/enthought/pyface/blob/4bccc3e4bba7b7748025b21e08359f446a2abaa7/pyface/ui/qt4/code_editor/code_widget.py#L39

And potentially, if the import fails we can just continue on with highlighter set to the default, which means graceful degradation of capabilities.

Something similar would be needed for the shell, with a slight complication that it interacts more with the highlighter.

There may be alternative approaches that work, but it feels like we want to handle this in the qt-specific code.

**Comment by corranwebster** [comment]
This change might also cause problems for TraitsUI because:

- the wx `CodeEditor` imports from `pyface.api` so it will fail it import time if Pygments is not installed, but Pygments isn't a requirement: https://github.com/enthought/traitsui/blob/01db38d6bce242de1426b279d38d5025fcbbfc32/traitsui/wx/code_editor.py#L35
- the qt `CodeEditor` directly imports from the Pyface internals, so it will still fail without Pygments: https://github.com/enthought/traitsui/blob/01db38d6bce242de1426b279d38d5025fcbbfc32/traitsui/qt4/code_editor.py#L20
- the generic `ShellEditor` will still fail at instantiation time because it doesn't import from pyface.api: https://github.com/enthought/traitsui/blob/148457a08b4f0fc14ccff6dc2166fa8f0251643c/traitsui/editors/shell_editor.py#L51

**Comment by kitchoi** [comment]
The concern for wx CodeEditor is legit.

> the qt CodeEditor directly imports from the Pyface internals, so it will still fail without Pygments:
> the generic ShellEditor will still fail at instantiation time because it doesn't import from pyface.api:

I think when one does need the Qt implementation of CodeEditor, then they need to have pygments. The goal here is not to make CodeEditor work without pygments. That could be a separate issue, though.

**Comment by kitchoi** [comment]
@corranwebster  I looked into the Qt specific code you mentioned, and I find that somewhat messy. While moving the import down in PythonShell could defer the import error, it defers the error all the way to when the class is instantiated.  The PythonShell, like you said, is even trickier because it is subclassing. And even if I try to make that work, the final solution does not look like it would last: It is too easy to add an import in the Qt specific modules that could reintroduce the dependency.

Taking a step back, at the moment, pygments is still a core dependency for PythonShell and PythonEditor, so failing early when they are imported seems fair. What we are trying to achieve it to make `api.py` more flexible.

I attempted the other approach which only touches `api.py`, which would preserve the import of PythonShell and PythonEditor for wx, and added more tests to ensure that.

**Comment by kitchoi** [comment]
#776 strikes again! :grumpy:

**Review by corranwebster — APPROVED** [review]
I'm OK with this implementation - it solves the immediate problem - however I think we will end up with different error reports from users about `PythonEditor` and `PythonShell` not importing correctly.  Time will tell if this is better or worse than the current situation.

**Comment by corranwebster** [comment]
I have an alternative suggestion:

- do the import error check in `pyface.python_shell` and `pyface.python_editor`
- if it fails, return an object which fails on instantiation with a message about needing `pygments`

**Comment by corranwebster** [comment]
Something like:
```
from .toolkit import toolkit_object

try:
    PythonEditor = toolkit_object("python_editor:PythonEditor")
except ImportError as _exception:
    # Excuse pygments dependency (for Qt), otherwise re-raise
    if _exception.name != "pygments":
        raise
    def PythonEditor(**traits):
        raise RuntimeError("PythonEditor on Qt requires Pygments to be installed")
```

**Comment by kitchoi** [comment]
Thanks @corranwebster 

> I have an alternative suggestion:
> 
> do the import error check in pyface.python_shell and pyface.python_editor
> if it fails, return an object which fails on instantiation with a message about needing pygments

Shall I open a separate issue for this implementation option and merge this one?

**Comment by kitchoi** [comment]
I also thought about raising a warning in the `api.py` but I thought that might be too noisy... We could however log at the debug level where we capture and silence the ImportError, which should be invisible by default. What do you think about that?

**Comment by kitchoi** [comment]
I will merge this now. I opened #800 for considering the alternative approach in the future (subject to feedback). I will open a separate PR on the logging idea.

**Comment by kitchoi** [comment]
For reference, the decision for not emitting a warning is partly similar to https://github.com/enthought/traitsui/issues/731. In TraitsUI, ArrayEditor is not available when NumPy is not available. While `traitsui.api` does not warn, `traitsui.editors.api` does. When one does not need the ArrayEditor, the import warning is just noise.

**Pull request merged**